### PR TITLE
cmake remove circular linking and reorganize

### PR DESCRIPTION
**Pull request by dagar**
PX4 is supposed to be modular, but because of the current cmake build system everything is built and linked together as a large pile of code. The lines between different modules and different layers of the system have become blurred in many cases.

This PR changes the PX4 cmake build slightly to allow us to actually enforce and maintain modularity. We can then tease apart some of these interconnected components and work towards more well defined layers (https://github.com/PX4/Firmware/pull/8784#issue-166153775). 

First instead of calling everything a module (px4_add_module) and linking it all together (with multiple passes to resolve circular dependencies) I changed px4_add_module to require a MAIN. Any modules included in a configuration (eg nuttx_px4fmu-v2_default.cmake) are included in the final PX4 binary. Modules without a main are now simply libraries (px4_add_module -> px4_add_library). Some libraries are part of PX4 and need access to uORB, the parameter system, logging system, etc (PX4 libraries: px4_add_library), but many don't (regular cmake: add_library). 

Libraries aren't automatically included (linked) in your final PX4 binary, they must be added as a module dependency. As a result we don't need to list them per configuration. Including a particular module pulls in a library only if it's needed as a dependency. With that dependency comes additional include directories and build flags if specified by the library.

## PX4 CMake Changes

**px4_add_module (PX4 Modules)**
 - now requires MAIN
 - access to uORB, params, and other libraries
 - name is defined for the logging system (PX4_INFO, PX4_WARN, PX4_ERR)
 - no access to other modules by default

**px4_add_library (PX4 libraries)**
 - access to PX4 uORB, params, etc
 - not linked to final binary by default, must be a dependency (DEPENDS) of a module
 - name is defined for the logging system (PX4_INFO, PX4_WARN, PX4_ERR)
 - use INTERFACE to specify include directories, build flags, etc that should be used by the module
 - can use other libraries (PX4 or not)

**add_library**
 - use for regular libraries that shouldn't have access to anything PX4 specific (uORB, params, etc)
 - can't use logging system (PX4_INFO, PX4_WARN, PX4_ERR)
 - not linked to final binary by default, must be a dependency (DEPENDS) of a module

**cmake configurations**
 - all libraries (src/lib) are automatically included in every build
 - only the ones that are dependencies of PX4 modules (px4_add_module DEPENDS) are actually built
 - other common pieces (eg src/drivers/boards/${BOARD}) are also included automatically now
 - this removes the need for every cmake configurations to exhaustively list every component and core system piece. Modules are specified directly, and then the libraries are pulled in as dependencies.

 - example, here's how the px4fmu-v2 configuration (nuttx_px4fmu-v2_default.cmake) is simplified.

   - image

This also removes the need to link with `--start-group/--end-group` to resolve the circular dependencies. 

## Notes
 * A few ideas taken from [CppCon 2017: Mathieu Ropert “Using Modern CMake Patterns to Enforce a Good Modular Design”](https://www.youtube.com/watch?v=eC9-iRN2b04).
   - declare all properties per target (PRIVATE, PUBLIC, or INTERFACE)
     - target_link_libraries for dependencies (somewhat poorly named, but very useful)
     - target_compile_defintions
     - target_compile_options
     - target_include_directories
   - forbid/report circular and hidden dependencies
   - build interfaces are transitive
   - help developer reason at module level

 * Related issues and PRs https://github.com/PX4/Firmware/pull/8784#issue-166153775, https://github.com/PX4/Firmware/issues/8127, https://github.com/PX4/Firmware/issues/7705, https://github.com/PX4/Firmware/pull/7806, and https://github.com/PX4/Firmware/issues/7318

### CODE REVIEW

**Comment by dagar** [comment]
@Stifael you might be interested in some of this. Once this is in PX4 we can update ECL with something more modular that's also a proper standalone cmake library.

File: src/systemcmds/tune_control/CMakeLists.txt [hunk 1]
```diff
@@ -38,6 +38,5 @@ px4_add_module(
 	SRCS
 		tune_control.cpp
 	DEPENDS
-		platforms__common
+		tunes
```
Comment (Stifael) [inline src/systemcmds/tune_control/CMakeLists.txt]: so much better

File: src/modules/simulator/CMakeLists.txt [hunk 2]
```diff
@@ -53,6 +53,8 @@ if (NOT ${OS} STREQUAL "qurt")
 		simulator_mavlink.cpp)
 endif()
 
+add_subdirectory(ledsim)
+
 px4_add_module(
```
Comment (Stifael) [inline src/modules/simulator/CMakeLists.txt]: would be nice to free ecl a from px4 dependencies such that it could be included similar to the simulator

**Comment by Stifael** [comment]
this makes the px4 cmake structure much more understandable

**Comment by Stifael** [comment]
this makes the px4 cmake structure much more understandable

**Comment by dagar** [comment]
Here's another good summary of modern cmake. https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1

File: src/modules/mc_pos_control/CMakeLists.txt [hunk 3]
```diff
@@ -40,6 +40,8 @@ px4_add_module(
 		PositionControl.cpp
 		Utility/ControlMath.cpp
 	DEPENDS
-		platforms__common
-	)
-# vim: set noet ft=cmake fenc=utf-8 ff=unix :
+		controllib
+		flight_tasks
+		git_ecl
```
Comment (bkueng) [inline src/modules/mc_pos_control/CMakeLists.txt]: Are both git_ecl and ecl needed?

Comment (dagar) [reply src/modules/mc_pos_control/CMakeLists.txt]: At the moment yes. The git_ecl target is on the Firmware side and ensures the submodule is actually in place. Then ecl is the actual library in the ecl project. Mixing submodule dependencies into the build system like this is a bit unusual, but has worked surprisingly well. 

Later we can split the ecl project into separate libs for ekf, tecs, etc that build standalone or within PX4.

**Comment by dagar** [comment]
Thanks for the review. Can you open an issue about any incremental build issues you hit?

**Comment by bkueng** [comment]
> Can you open an issue about any incremental build issues you hit?

#9369

File: src/modules/mc_pos_control/CMakeLists.txt (hunk 3 above)
Comment (bkueng) [reply src/modules/mc_pos_control/CMakeLists.txt]: Ok, it's just not obvious for someone adding a new module

**Pull request merged**